### PR TITLE
dist/tools/ethos: Build escaped packet in buffer before writing to UART

### PR DESCRIPTION
**Pull request by gebart**
Significantly improves throughput and latency on FRDM-KW41Z.

Ping against fe80::2 (ethos interface link local address) with a packet of size 100 yields a 30 ms roundtrip with this patch, 177 ms without. Size 1000 ping before: 1160 ms, after: 188 ms

### CODE REVIEW

**Comment by gebart** [comment]
Somehow the effect on the relayed packets to a remote gnrc_networking node is even greater.
Roundtrip comparison

without this PR:

```
PING fdfd::2107:5329:36b7:6e26(fdfd::2107:5329:36b7:6e26) 56 data bytes
64 bytes from fdfd::2107:5329:36b7:6e26: icmp_seq=1 ttl=64 time=1570 ms
64 bytes from fdfd::2107:5329:36b7:6e26: icmp_seq=2 ttl=64 time=629 ms
64 bytes from fdfd::2107:5329:36b7:6e26: icmp_seq=3 ttl=64 time=614 ms
64 bytes from fdfd::2107:5329:36b7:6e26: icmp_seq=4 ttl=64 time=613 ms
64 bytes from fdfd::2107:5329:36b7:6e26: icmp_seq=5 ttl=64 time=617 ms
^C
--- fdfd::2107:5329:36b7:6e26 ping statistics ---
6 packets transmitted, 5 received, 16% packet loss, time 5041ms
rtt min/avg/max/mdev = 613.833/809.254/1570.644/380.738 ms, pipe 2
```

with this PR:

```
PING fdfd::2107:5329:36b7:6e26(fdfd::2107:5329:36b7:6e26) 56 data bytes
64 bytes from fdfd::2107:5329:36b7:6e26: icmp_seq=1 ttl=64 time=65.1 ms
64 bytes from fdfd::2107:5329:36b7:6e26: icmp_seq=2 ttl=64 time=33.4 ms
64 bytes from fdfd::2107:5329:36b7:6e26: icmp_seq=3 ttl=64 time=32.0 ms
64 bytes from fdfd::2107:5329:36b7:6e26: icmp_seq=4 ttl=64 time=35.4 ms
64 bytes from fdfd::2107:5329:36b7:6e26: icmp_seq=5 ttl=64 time=34.1 ms
^C
--- fdfd::2107:5329:36b7:6e26 ping statistics ---
5 packets transmitted, 5 received, 0% packet loss, time 4004ms
rtt min/avg/max/mdev = 32.017/40.014/65.102/12.593 ms
```

The only change between the two runs was a recompile and restart of the ethos linux binary

File: dist/tools/ethos/ethos.c [hunk 1]
```diff
@@ -241,28 +254,41 @@ static int _serial_handle_byte(serial_t *serial, char c)
...
+            out[buffered++] = LINE_ESC_CHAR;
+            c ^= 0x20;
+            ++escaped;
+            if (buffered >= SERIAL_BUFFER_SIZE) {
+                checked_write(fd, out, buffered);
+                buffered = 0;
+            }
+        }
+        out[buffered++] = c;
+        if (buffered >= SERIAL_BUFFER_SIZE) {
```
Comment (kaspar030) [inline dist/tools/ethos/ethos.c]: If out contains more than SERIAL_BUFFER_SIZE bytes, didn't we write past it's size?

File: dist/tools/ethos/ethos.c [hunk 2]
```diff
@@ -49,8 +52,18 @@ static void usage(void)
 
 static void checked_write(int handle, void *buffer, int nbyte)
 {
-    if (write(handle, buffer, nbyte) != nbyte) {
-        fprintf(stderr, "write to fd %i failed: %s\n", handle, strerror(errno));
+    ssize_t res;
+    while (nbyte > 0) {
+        res = write(handle, buffer, nbyte);
+        if (res < 0) {
```
Comment (kaspar030) [inline dist/tools/ethos/ethos.c]: why two seperate if-cases? looks to me like one error message should be enough, with "res=%i".

**Comment by kaspar030** [comment]
Thanks for tackling this!

File: dist/tools/ethos/ethos.c (hunk 2 above)
Comment (gebart) [inline dist/tools/ethos/ethos.c]: Accidentally left over from debugging. I was afraid that the write would return 0 sometimes if the device driver's buffer is full, but it seems I was wrong, I have not seen a zero write happen yet.

File: dist/tools/ethos/ethos.c (hunk 1 above)
Comment (gebart) [inline dist/tools/ethos/ethos.c]: do you mean that it should only be necessary to check for equality? (`if (buffered == SERIAL_BUFFER_SIZE)`)

Comment (kaspar030) [inline dist/tools/ethos/ethos.c]: Hm, I just tried to find a case where that could happen. ```>=``` is probably safer.

**Comment by gebart** [comment]
@kaspar030 addressed comments

File: dist/tools/ethos/ethos.c [hunk 3]
```diff
@@ -241,27 +250,40 @@ static int _serial_handle_byte(serial_t *serial, char c)
...
-                escaped++;
-                break;
-            default:
-                checked_write(fd, &c, 1);
+        unsigned char c = *buf++;
+        if (c == LINE_FRAME_DELIMITER || c == LINE_ESC_CHAR) {
+            out[buffered++] = LINE_ESC_CHAR;
+            c ^= 0x20;
+            ++escaped;
+            if (buffered >= SERIAL_BUFFER_SIZE) {
```
Comment (kaspar030) [inline dist/tools/ethos/ethos.c]: would it make sense to factor out these three lines into a static function?
```
static void _maybe_write(fd, out, *buffered, threshold)
{
  if (*buffered >= threshold) {
     checked_write(fd, out, *buffered);
     *buffered = 0;
  }
}
```

Could be used three times.

Comment (smlng) [inline dist/tools/ethos/ethos.c]: or alternatively extend the already existing `checked_write`function

Comment (kaspar030) [inline dist/tools/ethos/ethos.c]: Good idea! @gebart feel free to do so or just squash and we do that as folo.

**Comment by kaspar030** [comment]
Quickly tested gnrc_border_router with samr21-xpro, still works.

File: dist/tools/ethos/ethos.c (hunk 3 above)
Comment (gebart) [inline dist/tools/ethos/ethos.c]: @kaspar030 let's do it as a follow-up

**Comment by gebart** [comment]
squashed

**Pull request merged**